### DISCUSSION TITLE
buildpack for static websites

### DISCUSSION THREAD

**Discussion by r6m** [paketo-buildpacks/feedback#14] [Q&A] [answered]
I'm new to buildpacks world and want some help if it's the right choice for me.
Is there a buildpack for react-app or hugo which I could pass the build directory to the nginx for just serving static files?

**Comment by dmikusa** [comment, marked as answer]
The [Paketo Staticfile buildpack](https://github.com/paketo-community/staticfile) targets this use case. It is still in development, but you can give it a try and provide feedback through issues on that repo.

Alternatively, you could provide your own [nginx conf](https://github.com/paketo-buildpacks/nginx) or [Apache HTTPD conf](https://github.com/paketo-buildpacks/httpd) and use the server directly. This requires a little more upfront work than using Staticfile, but allows you to customize things to a degree you cannot with the Staticfile buildpack.

Hope that helps!

**Reply by r6m** [reply]
Thanks @dmikusa-pivotal.

**Reply by r6m** [reply]
Is there any example repo to check?
BTW I couldn't start a NextJS project using paketo, but heroku buildpacks worked.
Here are buildpacks found for a NextJS project:
```bash
5 of 9 buildpacks participating
paketo-buildpacks/ca-certificates 2.3.2
paketo-buildpacks/node-engine     0.6.2
paketo-buildpacks/yarn            0.4.0
paketo-buildpacks/yarn-install    0.4.0
paketo-buildpacks/yarn-start      0.2.0
```
NextJS project must be built before start using `yarn build (next build)` command. how can I add/contribute `yarn-build` as a buildpack?

**Reply by dmikusa** [reply]
My suggestions above would be for passing in the final static build of your project. I thought that was what you were looking for when you mentioned `I could pass the build directory`.

I don't know NextJS, but when I use this for my Vue.js projects, I do a `npm run build` which gives me a `dist/` folder. I would then `pack build -p dist/` which passes the contents of the `dist/` folder to the buildpacks. The staticfile buildpack can run that because `dist/` is just static HTML/CSS/JS.

If you are trying to use buildpacks to build a JS project, something that would require NPM/Yarn, you would not be able to use the Staticfile buildpack. That buildpack has no build tools, just web servers. You'd have to use the Node.js buildpack, or something that provides your build tools. 

I don't know the latest on Node.js support or if there's a way to customize what you're asking. I believe @arjun024 is more familiar so he might be able to give you some updates if you're trying to have the buildpacks actually build the project. I'll say that because most JS projects don't have any native code, they are pretty easy to build on your local laptop and use Staticfile for deployment.

**Reply by fg-j** [reply]
Perhaps the angular, vue, and react examples in [the Paketo samples repo](https://github.com/paketo-buildpacks/samples/tree/main/nodejs) could point you in a helpful direction, @r6m. None of those use NGINX to serve the app, but you could probably use the Nodejs and NGINX buildpacks together to accomplish that goal.

**Reply by ryanmoran** [reply]
One of the key points in building a static JS app is that you will need to run some sort of build process that will generate your static files. The [`node-run-script` buildpack](https://github.com/paketo-buildpacks/node-run-script) is what supports this feature. In order to indicate to that buildpack what it should run, you need to set the `BP_NODE_RUN_SCRIPTS` environment variable as described [here](https://paketo.io/docs/howto/nodejs/#specifying-scripts-to-run-during-build-phase). If you do this, and then include the `staticfile`, `nginx`, or `httpd` buildpacks as the last buildpack in your ordering, then you can get the outcome you are looking for.

**Reply by arjun024** [reply]
Here's a react (for build) + nginx (for run) example if you wanna checkout. Nginx buildpack here overrides the start command set by the Node.js buildpack to start the server using the `nginx.conf` at the root of the app. https://github.com/arjun024/react-nginx

**Reply by fg-j** [reply]
Love this sample app! @arjun024 Would you be willing to contribute it to the Paketo samples repo? I imagine this is a common use case to show off.

**Reply by r6m** [reply]
Thanks @fg-j, @arjun024, @ryanmoran and @dmikusa-pivotal 
I built the react example and it did work well.
most of nodejs front-end and some backends now have a `build` script in `package.json`. maybe it could be simpler by just detecting `build` script in `package.json` and run build command instead of passing `BP_NODE_RUN_SCRIPTS` env variable. I mean we could just pass `gcr.io/paketo-buildpacks/nodejs` buildpack and if `package.json` has `build` script, then include `paketo-buildpacks/npm-run-script` or even something like `paketo-buildpacks/yarn-build`. This is how `paketo-buildpacks/yarn-start` is added. :)
